broker.go

```diff
@@ -0,0 +1,606 @@
+}
+
+// SetAllowPeer configures Broker with a function to evaluate whether to allow
+// or reject messages from the specified peer.  Setting nil removes any
+// filtering and allows messages from all peers.  This replaces any previously
+// configured AllowPeerFunc.
+func (lb *Broker) SetAllowPeer(allowPeer AllowPeerFunc) {
+	lb.handlersMutex.Lock()
+	defer lb.handlersMutex.Unlock()
+	lb.allowPeer = allowPeer
+}
+
+// Close shuts down the broker.
+func (lb *Broker) Close() error {
+	var err error
+	lb.closeOnce.Do(func() {
+		err = lb.doClose()
+	})
+	return err
+}
+
+func (lb *Broker) doClose() error {
+	lb.psub.Cancel()
+
+	var err, errs error
+	if err = lb.dtSync.Close(); err != nil {
+		errs = multierror.Append(errs, err)
+	}
+
+	// If Broker owns the pubsub topic, then close it.
+	if lb.topic != nil {
+		if err = lb.topic.Close(); err != nil {
+			log.Errorf("Failed to close pubsub topic: %s", err)
+			errs = multierror.Append(errs, err)
+		}
+	}
+
+	// Dismiss any event readers.
+	lb.outEventsMutex.Lock()
+	for _, ch := range lb.outEventsChans {
+		close(ch)
+	}
+	lb.outEventsChans = nil
+	lb.outEventsMutex.Unlock()
+
+	// Shutdown pubsub services.
+	lb.cancelps()
+
+	// Stop the distribution goroutine.
+	close(lb.inEvents)
+	return errs
+}
+
+// OnSyncFinished creates a channel that receives change notifications, and
+// adds that channel to the list of notification channels.
+//
+// Calling the returned cancel function removed the notification channel from
+// those the receive change notifications, and closes the channel to allow any
+// waiting goroutines to stop waiting on the channel.  For this reason, channel
+// readers should check for channel closure.
+func (lb *Broker) OnSyncFinished() (<-chan SyncFinished, context.CancelFunc) {
+	// Channel is buffered to prevent distribute() from blocking if a reader is
+	// not reading the channel immediately.
+	ch := make(chan SyncFinished, 1)
+	lb.outEventsMutex.Lock()
+	defer lb.outEventsMutex.Unlock()
+
+	lb.outEventsChans = append(lb.outEventsChans, ch)
+	cncl := func() {
+		lb.outEventsMutex.Lock()
+		defer lb.outEventsMutex.Unlock()
+		for i, ca := range lb.outEventsChans {
+			if ca == ch {
+				lb.outEventsChans[i] = lb.outEventsChans[len(lb.outEventsChans)-1]
+				lb.outEventsChans[len(lb.outEventsChans)-1] = nil
+				lb.outEventsChans = lb.outEventsChans[:len(lb.outEventsChans)-1]
+				close(ch)
+				break
+			}
+		}
+	}
+	return ch, cncl
+}
+
+// Sync performs a one-off explicit sync with the given peer for a specific CID
+// and updates the latest synced link to it.
+//
+// The Context passed in controls the lifetime of the background sync process,
+// so make sure the context has an appropriate deadline, if any.
+//
+// A SyncFinished channel is returned to allow the caller to wait for this
+// particular sync to complete.  Any OnSyncFinished readers will also get a
+// SyncFinished when the sync succeeds, but only if syncing to the latest using
+// the default selector.
```

Comment:
```suggestion
// the default selector and a `cid.Undef`.
```

But maybe this is specified by saying "syncing to the latest." so feel free to disregard if you think it's clear enough as is.